### ISSUE TITLE
issues running on azure app service

### ISSUE THREAD

**Issue by jamesdixon** [lovell/sharp#631]
Hi,

First off, thanks for the great work on Sharp! 

I'm trying to use it within a webjob that I run on Azure App Service. I actually do all of the compilation through the Appveyor CI service and then transfer those files to Azure. Whenever I attempt to run the job that uses Sharp, I receive the following error:

```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException: Microsoft.Azure.WebJobs.Host.FunctionInvocationException: Exception while executing function: Functions.ProcessImages ---> System.Exception: Error: A dynamic link library (DLL) initialization routine failed.
 \\?\D:\local\Temp\jobs\continuous\scripthost\mvhbfltp.51e\ProcessImages\node_modules\sharp\build\Release\sharp.node at Error (native) at Object.Module._extensions..node (module.js:583:18) at Module.load (module.js:473:32) at tryModuleLoad (module.js:432:12) at Function.Module._load (module.js:424:3) at Module.require (module.js:483:17) at require (internal/module.js:20:19) at Object.<anonymous> (D:\local\Temp\jobs\continuous\scripthost\mvhbfltp.51e\ProcessImages\node_modules\sharp\index.js:12:13) at Module._compile (module.js:556:32) at Object.Module._extensions..js (module.js:565:10) at Module.load (module.js:473:32) at tryModuleLoad (module.js:432:12) at Function.Module._load (module.js:424:3) at Module.require (module.js:483:17) at require (internal/module.js:20:19) at Object.<anonymous> (D:\local\Temp\jobs\continuous\scripthost\mvhbfltp.51e\ProcessImages\index.js:11:15) at Module._compile (module.js:556:32) at Object.Module._extensions..js (module.js:565:10) at Module.load (module.js:473:32) at tryModuleLoad (module.js:432:12) at Function.Module._load (module.js:424:3) at Module.require (module.js:483:17)
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at Microsoft.Azure.WebJobs.Script.Description.FunctionInvokerBase.<Invoke>d__23.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`1.<InvokeAsync>d__0.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<InvokeAsync>d__53.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithWatchersAsync>d__3b.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__36.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__1a.MoveNext()
 --- End of inner exception stack trace ---
 at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__1a.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<TryExecuteAsync>d__2.MoveNext()
```

I'm running Node 6.9.1 x64 on Azure. When I compile the application on Appveyor, I'm also using Node 6.9.1 x64. Compilation works great, it's just the error above at runtime.

Appreciate any help you can provide.

Cheers!

**Comment by lovell** [comment]
Hello, the most salient part of this error message is `A dynamic link library (DLL) initialization routine failed`.

> "and then transfer those files to Azure"

Are all of the `.dll` and `.lib` files that are copied to become siblings of `build/Release/sharp.node` during the `npm install` phase included in the list of transferred files?

**Comment by jamesdixon** [comment]
Hello!

Thanks for the quick reply. I just double-checked and yes, there are a bunch of `.lib` and `.dll` files under `node_modules/sharp/build/Release`.

Any other thoughts on things to check?

Thank you again for your assistance!

Best,
James

**Comment by lovell** [comment]
[Dependency Walker](http://www.dependencywalker.com/) is my usual tool of choice to help troubleshoot DLL linking problems. Good luck!

**Comment by lovell** [comment]
@jamesdixon Were you able to make any progress with Dependency Walker (or similar) on the Azure instance?

**Comment by jamesdixon** [comment]
@lovell I apologize for not responding!

I did some testing and I do not believe it's anything to do with Sharp. In case someone else runs into this, it appears that there are issues with native Node binaries when using them within Azure Webjobs that are running on Azure App Service. The library works as expected when using it within an application running on Azure App Service, just not within a webjob. I've opened an issue in the [Azure Webjob repo](https://github.com/Azure/azure-webjobs-sdk-script/issues/989).

Thank you!
